homeassistant/components/tuya/__init__.py

```diff
@@ -191,31 +289,88 @@ async def async_unload_entry(hass, entry):
         )
     )
     if unload_ok:
+        for listener in hass.data[DOMAIN][entry.entry_id]["listener"]:
+            listener()
+        hass.data[DOMAIN].pop(entry.entry_id)
         hass.data[DOMAIN][ENTRY_IS_SETUP] = set()
         hass.data[DOMAIN][TUYA_TRACKER]()
         hass.data[DOMAIN][TUYA_TRACKER] = None
         hass.data[DOMAIN][TUYA_DATA] = None
+        hass.data[DOMAIN].pop(TUYA_DEVICES_CONF)
         hass.services.async_remove(DOMAIN, SERVICE_FORCE_UPDATE)
         hass.services.async_remove(DOMAIN, SERVICE_PULL_DEVICES)
         hass.data.pop(DOMAIN)
 
     return unload_ok
 
 
+async def update_listener(hass, entry):
+    """Update when config_entry options update."""
+    _update_discovery_interval(
+        hass, entry.options.get(CONF_DISCOVERY_INTERVAL, DEFAULT_DISCOVERY_INTERVAL)
+    )
+    _update_query_interval(
+        hass, entry.options.get(CONF_QUERY_INTERVAL, DEFAULT_QUERY_INTERVAL)
+    )
+
+
+async def cleanup_device_registry(hass, device_id):
+    """Remove device registry entry if there are no remaining entities."""
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    entity_registry = await hass.helpers.entity_registry.async_get_registry()
+    if device_id and not hass.helpers.entity_registry.async_entries_for_device(
+        entity_registry, device_id
+    ):
+        device_registry.async_remove_device(device_id)
+
+
 class TuyaDevice(Entity):
     """Tuya base device."""
 
+    _device_count = 0
+
     def __init__(self, tuya, platform):
         """Init Tuya devices."""
         self._tuya = tuya
         self._tuya_platform = platform
+        self._dev_conf = None
+
+    def _inc_device_count(self):
+        dev_type = self._tuya.device_type()
+        if any(c in dev_type for c in ["scene", "switch"]):
+            return
+        TuyaDevice._device_count += 1
+
+    def _dec_device_count(self):
+        dev_type = self._tuya.device_type()
+        if any(c in dev_type for c in ["scene", "switch"]):
+            return
+        if TuyaDevice._device_count > 0:
```

Comment:
Why is this check needed?

Reply:
Just a check to be sure that device count don't go below 0, but probably not really useful.

Reply:
It seems like it would be a bug if it decremented the device count too far.  Ideally we would have a test for this instead of a guard in the code.

Reply:
Can you better clarify what you mean?

Reply:
Removed check, unuseful